Credentials authentication fix

xsteelej opened the pull request:
<!-- 
Thank you for contributing to ODK Collect!

Before sending this PR, please read
https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md
-->

Closes #2541 

#### What has been done to verify that this works as intended?
Configured a local Aggregate VM and tested that Digest authentication works. I.e. it prompts for a username and password and records the credentials in shared preferences.

#### Why is this the best possible solution? Were any other approaches considered?
It fits in with the current design paradigms.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
No behaviour changes have been made, the bug fixes cases where the Anonymous user on the Aggregate server has switched off Data Collector, Data Viewer, Form Manager

#### Do we need any specific form for testing your changes? If so, please attach one.
No

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

xsteelej committed:
Update README.md

xsteelej committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	README.md

xsteelej committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Fixed a bug with credentials. Basically WebCredentialsUtils.getCredentials(...)  was not comparing against the host part of the Server Preferences Url. It meant that the username / password were not being correctly retrieved from preferences.

lognaturel commented:
Thanks! Can you please file a quick issue so it’s clear how to reproduce the bug and verify it is fully addressed?

lognaturel commented:
I can see that most of the changes are direct refactors with no impact on functionality. I see two functional changes:
1. `saveCredentials` now saves the credentials to shared preferences if the host matches the host in shared preferences (https://github.com/opendatakit/collect/pull/2539/files#diff-57ed72a193025ad0657b28dc6c86ec7cR31)

2. `getCredentials` now reads the credentials from shared preferences if the host matches the host in shared preferences (https://github.com/opendatakit/collect/pull/2539/files#diff-57ed72a193025ad0657b28dc6c86ec7cR85)

 I tried the following cases with Aggregate hosted on AppEngine and required credentials:
- URL only with no username and password; prompted for username and password on fetch and those are correctly saved in preferences
- URL with bad username and password; prompted for username and password on fetch and those are correctly saved in preferences

Is the problem you observed perhaps when there is a redirect?

xsteelej commented:
Hi @lognaturel,

Apologies for the delay, yes I'll add an issue that describes the problem.

You are correct with the two points you raised above. The main problem was that the code wasn't retrieving the credentials from preferences when the host matched the server preferences.

lognaturel commented:
> The main problem was that the code wasn't retrieving the credentials from preferences when the host matched the server preferences.

Ok! I wasn't able to reproduce with that info so if you can provide specific steps, preferably with publicly-accessible servers, that'd be great. If it has to do with your Aggregate setup, describing that would be helpful. Thanks!

xsteelej commented:
Hi @lognaturel - I've just managed to reproduce it following these steps:

**On Aggregate server**
1. Login as administrator
2. Select "Site Admin"
3. Untick Data Collector on "anonymousUser" (and any other columns that may be ticked)
4. Save Changes

**Collect**
1. Navigate to General Settings->Server
2. Enter the URL of the Server
3. Leave out the username / password

**Steps to reproduce**
1. After configuring as above navigate to "Get Blank Form"
2. Observe that the "Server Requires Authentication" pops up
3. Type in the administrator username / password (Or other username that has Data Collector ticked).
4. Observe that the request fails and Collect asks for the username / password over and over.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/AuthDialogUtility.java:
```diff
@@ -82,14 +84,7 @@ public void onClick(DialogInterface dialog, int which) {
-                if (customUsername != null && customPassword != null) {
-                    webCredentialsUtils.saveCredentials(finalOverriddenUrl != null ? finalOverriddenUrl : webCredentialsUtils.getServerFromPreferences(), userNameValue, passwordValue);
-                } else if (finalOverriddenUrl == null) {
-                    webCredentialsUtils.saveCredentialsPreferences(userNameValue, passwordValue);
```
This change here means that username and password are never saved when provided to the dialog, right? Is that intentional? Is it possible this change resulted in the issue you were seeing?

This somewhat convoluted block was added here because now it's possible to override the settings with intent extras (#2461). Those values should never be saved to preferences.

xsteelej replied:
Hi @lognaturel - I moved the decision about where to save the credentials into saveCredentials - if the host is the same as what is configured in the server then save to preferences, otherwise temporarily save to memory.

lognaturel replied:
This is perhaps not super common but it could be possible for the settings to have server A configured without username and password and then for an external application to set temporary credentials to fetch or send forms *from server A*. In that case, the credentials would now be saved in Collect's preferences. Did I get that right? If so, that's a change in behavior -- previously credentials from an external intent would never be saved.

xsteelej replied:
Ok, i'll check that

lognaturel commented:
From #2541: `when the URL has a port attached to the host then the problem occurs.` This can be reproduced by setting the server to an AppEngine instance and appending a port number (e.g. https://nafundi-test.appspot.com:443)

I can confirm his is a good change, then. The only issue I see is the question about intent overrides I brought up [here](https://github.com/opendatakit/collect/pull/2539#discussion_r218243472).

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into authentication_fix

* master:
  Update September beta translations
  Implemented RecyclerView to improve selectOne speed (#2513)
  Replace download service with activity (#2463)
  Simplify gradle commands to increase efficiency (#2530)

# Conflicts:
#	collect_app/src/main/java/org/odk/collect/android/utilities/WebCredentialsUtils.java

lognaturel commented:
Since this is a bug fix, we don't need to rush to get this in before code freeze in 24ish hours. It would be good to get it to QA by Thursday, though. @xsteelej, if you don't have a chance to revert the part of the change that saves credentials that come from external intent extras before then, I can do it!

xsteelej commented:
Hi @lognaturel, I'll revert the change now.

jrsteele09 committed:
- Reverted AuthDialogUtility so that the case where Collect is launched from an external app, that supplies the username / password, then the credentials are not saved to preferences
- Updated WebCredentials so that saveCredentials does not save to preferences if the URL host matches the preferences server host
- Updated WebCredentials.getCredentials so that it returns the temporary username / password for the server preferences if the host has an entry in the temporary store

lognaturel commented:
Thank you, this looks good to me! For QA, the most important things to verify that I can think of are that 1) if the server address has a port number, credentials are still correctly fetched. 2) if the settings have a particular server set and an external app launches a form fetch or instance upload with the same server URL and a url and password, that those credentials are not saved.

mmarciniak90 commented:
@lognaturel 
I was able to reproduce the problem on my testing Aggregate server.
Credentials are saved after the fix.

I tested cases with the external app:
Case 1:
1. ODK Collect has server 'A' with a port number set. 
2. The external app has server 'B' with a port number set.   
3. The user tries to download a form in the external app via FORM_ID
4. The user needs to add username and password every time when tries to download a form -> credentials are not saved

Case 2:
1. ODK Collect has server 'A' with a port number set. 
2. The external app has server 'A' with a port set.   
3. The user tries to download a form in the external app via FORM_ID
4. The user needs to add username and password 
5. Credentials are saved in ODK Collect and the user does not need to add username and password when using the external app

If both results are accepted, the issue is ready for merge.

Verified on Android 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

ekigamba commented:
Nice work @xsteelej 😄 

I have a few comments, so as @lognaturel said [here](https://github.com/opendatakit/collect/pull/2539#discussion_r218243472), the values entered during intent overrides should never be saved. I also dunno why the previous refactor did not cover saving to preferences instead of just saving the new preferences to memory

grzesiek2010 commented:
@mmarciniak90 thanks for testing and pointing out those cases. I agree with @ekigamba that an external app shouldn't override our credentials. I'm sure @lognaturel is going to agree too.

lognaturel commented:
@ekigamba The normal case with an external app -- correct credentials for a URL -- is well-supported by this. That is, even if the URL matches the URL in the Collect preferences, the credentials aren't saved.

Case 2 that @mmarciniak90 describes is really unusual so I don't think it's a big deal. Note that in that case the server has a port number which generally only applies to local Aggregate installs. But do let us know if you think otherwise!

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
